// @binance-chain/javascript-sdk/lib/rpc/client.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _big = _interopRequireDefault(require("big.js"));

var _types = require("../decoder/types");

var decoder = _interopRequireWildcard(require("../decoder"));

var crypto = _interopRequireWildcard(require("../crypto"));

var _ = _interopRequireDefault(require("."));

var _validateHelper = require("../utils/validateHelper");

/**
 * @module rpc
 */
var BASENUMBER = Math.pow(10, 8);

var divide = function divide(num) {
  return +new _big.default(num).div(BASENUMBER).toString();
};

var convertObjectArrayNum = function convertObjectArrayNum(objArr, keys) {
  objArr.forEach(function (item) {
    keys.forEach(function (key) {
      item[key] = divide(item[key]);
    });
  });
};
/**
 * The Binance Chain Node rpc client
 */


var Client =
/*#__PURE__*/
function (_BaseRpc) {
  (0, _inherits2.default)(Client, _BaseRpc);

  /**
   * @param {String} uriString dataseed address
   * @param {String} netWork Binance Chain network
   */
  function Client() {
    var _this;

    var uriString = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "localhost:27146";
    var netWork = arguments.length > 1 ? arguments[1] : undefined;
    (0, _classCallCheck2.default)(this, Client);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Client).call(this, uriString));
    _this.netWork = netWork || "mainnet";
    return _this;
  }

  (0, _createClass2.default)(Client, [{
    key: "getBech32Prefix",
    value: function getBech32Prefix() {
      if (this.netWork === "mainnet") {
        return "bnb";
      }

      if (this.netWork === "testnet") {
        return "tbnb";
      }

      return "";
    }
    /**
     * @param {String} symbol - required
     * @returns {Object} token detail info
     */

  }, {
    key: "getTokenInfo",
    value: function () {
      var _getTokenInfo = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(symbol) {
        var path, res, bytes, result, _decoder$unMarshalBin, tokenInfo, bech32Prefix;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                (0, _validateHelper.validateSymbol)(symbol);
                path = "/tokens/info/" + symbol;
                _context.next = 4;
                return this.abciQuery({
                  path: path
                });

              case 4:
                res = _context.sent;
                bytes = Buffer.from(res.response.value, "base64");
                result = new _types.Token();
                _decoder$unMarshalBin = decoder.unMarshalBinaryLengthPrefixed(bytes, result), tokenInfo = _decoder$unMarshalBin.val;
                bech32Prefix = this.getBech32Prefix();
                tokenInfo.owner = crypto.encodeAddress(tokenInfo.owner, bech32Prefix);
                return _context.abrupt("return", tokenInfo);

              case 11:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getTokenInfo(_x) {
        return _getTokenInfo.apply(this, arguments);
      }

      return getTokenInfo;
    }()
    /**
     * get tokens by offset and limit
     * @param {Number} offset
     * @param {Number} limit
     * @returns {Array} token list
     */

  }, {
    key: "listAllTokens",
    value: function () {
      var _listAllTokens = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(offset, limit) {
        var path, res, bytes, result, _decoder$unMarshalBin2, tokenList, bech32Prefix;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                (0, _validateHelper.validateOffsetLimit)(offset, limit);
                path = "tokens/list/".concat(offset, "/").concat(limit);
                _context2.next = 4;
                return this.abciQuery({
                  path: path
                });

              case 4:
                res = _context2.sent;
                bytes = Buffer.from(res.response.value, "base64");
                result = [new _types.TokenOfList()];
                _decoder$unMarshalBin2 = decoder.unMarshalBinaryLengthPrefixed(bytes, result), tokenList = _decoder$unMarshalBin2.val;
                bech32Prefix = this.getBech32Prefix();
                tokenList.forEach(function (item) {
                  item.owner = crypto.encodeAddress(item.owner, bech32Prefix);
                });
                return _context2.abrupt("return", tokenList);

              case 11:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function listAllTokens(_x2, _x3) {
        return _listAllTokens.apply(this, arguments);
      }

      return listAllTokens;
    }()
    /**
     * @param {String} address
     * @returns {Object} Account info
     */

  }, {
    key: "getAccount",
    value: function () {
      var _getAccount = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(address) {
        var addr, addrHex, res, result, bytes, _decoder$unMarshalBin3, accountInfo, bech32Prefix;

        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                addr = crypto.decodeAddress(address);
                addrHex = Buffer.concat([Buffer.from("account:"), addr]);
                _context3.next = 4;
                return this.abciQuery({
                  path: "/store/acc/key",
                  data: addrHex
                });

              case 4:
                res = _context3.sent;
                result = new _types.AppAccount();
                bytes = Buffer.from(res.response.value, "base64");
                _decoder$unMarshalBin3 = decoder.unMarshalBinaryBare(bytes, result), accountInfo = _decoder$unMarshalBin3.val;
                bech32Prefix = this.getBech32Prefix();
                accountInfo.base.address = crypto.encodeAddress(accountInfo.base.address, bech32Prefix);
                return _context3.abrupt("return", accountInfo);

              case 11:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getAccount(_x4) {
        return _getAccount.apply(this, arguments);
      }

      return getAccount;
    }()
    /**
     * @param {Array} balances
     */

  }, {
    key: "getBalances",
    value: function () {
      var _getBalances = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(address) {
        var account, coins, balances;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.getAccount(address);

              case 2:
                account = _context4.sent;
                coins = [];
                balances = [];

                if (account) {
                  coins = account.base && account.base.coins || [];
                  convertObjectArrayNum(coins, ["amount"]);
                  convertObjectArrayNum(account.locked, ["amount"]);
                  convertObjectArrayNum(account.frozen, ["amount"]);
                }

                coins.forEach(function (item) {
                  var locked = account.locked.find(function (lockedItem) {
                    return item.denom === lockedItem.denom;
                  }) || {};
                  var frozen = account.frozen.find(function (frozenItem) {
                    return item.denom === frozenItem.denom;
                  }) || {};
                  var bal = new _types.TokenBalance();
                  bal.symbol = item.denom;
                  bal.free = item.amount;
                  bal.locked = locked.amount || 0;
                  bal.frozen = frozen.amount || 0;
                  balances.push(bal);
                });
                return _context4.abrupt("return", balances);

              case 8:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getBalances(_x5) {
        return _getBalances.apply(this, arguments);
      }

      return getBalances;
    }()
    /**
     * get balance by symbol and address
     * @param {String} address
     * @param {String} symbol
     * @returns {Object}
     */

  }, {
    key: "getBalance",
    value: function () {
      var _getBalance = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee5(address, symbol) {
        var balances, bal;
        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                (0, _validateHelper.validateSymbol)(symbol);
                _context5.next = 3;
                return this.getBalances(address);

              case 3:
                balances = _context5.sent;
                bal = balances.find(function (item) {
                  return item.symbol.toUpperCase() === symbol.toUpperCase();
                });
                return _context5.abrupt("return", bal);

              case 6:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getBalance(_x6, _x7) {
        return _getBalance.apply(this, arguments);
      }

      return getBalance;
    }()
    /**
     * @param {String} address
     * @param {String} symbol
     * @returns {Object}
     */

  }, {
    key: "getOpenOrders",
    value: function () {
      var _getOpenOrders = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee6(address, symbol) {
        var path, res, bytes, result, _decoder$unMarshalBin4, openOrders;

        return _regenerator.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                path = "/dex/openorders/".concat(symbol, "/").concat(address);
                _context6.next = 3;
                return this.abciQuery({
                  path: path
                });

              case 3:
                res = _context6.sent;
                bytes = Buffer.from(res.response.value, "base64");
                result = [new _types.OpenOrder()];
                _decoder$unMarshalBin4 = decoder.unMarshalBinaryLengthPrefixed(bytes, result), openOrders = _decoder$unMarshalBin4.val;
                convertObjectArrayNum(openOrders, ["price", "quantity", "cumQty"]);
                return _context6.abrupt("return", openOrders);

              case 9:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getOpenOrders(_x8, _x9) {
        return _getOpenOrders.apply(this, arguments);
      }

      return getOpenOrders;
    }()
    /**
     * @param {Number} offset
     * @param {Number} limit
     * @returns {Array}
     */

  }, {
    key: "getTradingPairs",
    value: function () {
      var _getTradingPairs = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee7(offset, limit) {
        var path, res, bytes, result, _decoder$unMarshalBin5, tradingPairs;

        return _regenerator.default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                (0, _validateHelper.validateOffsetLimit)(offset, limit);
                path = "/dex/pairs/".concat(offset, "/").concat(limit);
                _context7.next = 4;
                return this.abciQuery({
                  path: path
                });

              case 4:
                res = _context7.sent;
                bytes = Buffer.from(res.response.value, "base64");
                result = [new _types.TradingPair()];
                _decoder$unMarshalBin5 = decoder.unMarshalBinaryLengthPrefixed(bytes, result), tradingPairs = _decoder$unMarshalBin5.val;
                convertObjectArrayNum(tradingPairs, ["list_price", "tick_size", "lot_size"]);
                return _context7.abrupt("return", tradingPairs);

              case 10:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getTradingPairs(_x10, _x11) {
        return _getTradingPairs.apply(this, arguments);
      }

      return getTradingPairs;
    }()
    /**
     * @param {String} tradePair
     * @returns {Array}
     */

  }, {
    key: "getDepth",
    value: function () {
      var _getDepth = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee8(tradePair) {
        var path, res, bytes, result, _decoder$unMarshalBin6, depth;

        return _regenerator.default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                (0, _validateHelper.validateTradingPair)(tradePair);
                path = "dex/orderbook/".concat(tradePair);
                _context8.next = 4;
                return this.abciQuery({
                  path: path
                });

              case 4:
                res = _context8.sent;
                bytes = Buffer.from(res.response.value, "base64");
                result = new _types.OrderBook();
                _decoder$unMarshalBin6 = decoder.unMarshalBinaryLengthPrefixed(bytes, result), depth = _decoder$unMarshalBin6.val;
                convertObjectArrayNum(depth.levels, ["buyQty", "buyPrice", "sellQty", "sellPrice"]);
                return _context8.abrupt("return", depth);

              case 10:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getDepth(_x12) {
        return _getDepth.apply(this, arguments);
      }

      return getDepth;
    }()
  }]);
  return Client;
}(_.default);

var _default = Client;
exports.default = _default;
